pkg/gadgets/trace/dns/tracer/bpf/dns.c

```diff
@@ -114,159 +104,63 @@ struct {
-		if (skip != 0) {
-			skip--;
-		} else {
-			int label_len = load_byte(
-				skb, DNS_OFF + sizeof(struct dnshdr) + i);
-			if (label_len == 0)
-				break;
-			// The simple solution "i += label_len" gives verifier
-			// errors, so work around with skip.
-			skip = label_len;
-		}
-	}
+	struct event_t event;
+	__u16 sport, dport;
+	__u16 id;
+	int i;
 
-	return i < MAX_DNS_NAME ? i : MAX_DNS_NAME;
-}
+	// Skip non-IP packets
+	if (load_half(skb, offsetof(struct ethhdr, h_proto)) != ETH_P_IP)
+		return 0;
 
-// Save the IPv4 and IPv6 addresses in event->anaddr. Returns the number of saved addresses.
-static __always_inline int load_addresses(struct __sk_buff *skb, int ancount,
-					  int anoffset, struct event_t *event)
-{
-	int rroffset = anoffset;
-	int index = 0;
-	for (int i = 0; i < ancount && i < MAX_ADDR_ANSWERS; i++) {
-		__u16 rrname =
-			load_byte(skb, rroffset + offsetof(struct dnsrr, name));
-
-		// In most cases, the name will be compressed to two octets (indicated by first two bits 0b11).
-		// The offset calculations below assume compression, so exit early if the name isn't compressed.
-		if ((rrname & 0xf0) != 0xc0)
-			return 0;
-
-		// Safe to assume that all answers refer to the same domain name
-		// because we verified earlier that there's exactly one question.
-
-		__u16 rrtype =
-			load_half(skb, rroffset + offsetof(struct dnsrr, type));
-		__u16 rrclass = load_half(skb, rroffset + offsetof(struct dnsrr,
-								   class));
-		__u16 rdlength = load_half(
-			skb, rroffset + offsetof(struct dnsrr, rdlength));
-
-		if (rrtype == DNS_TYPE_A && rrclass == DNS_CLASS_IN &&
-		    rdlength == 4) {
-			// A record contains an IPv4 address.
-			// Encode this as IPv4-mapped-IPv6 in the BPF event (::ffff:<ipv4>)
-			// https://datatracker.ietf.org/doc/html/rfc4291#section-2.5.5.2
-			__builtin_memset(&event->anaddr[index][0], 0x0, 10);
-			__builtin_memset(&event->anaddr[index][10], 0xff, 2);
-			bpf_skb_load_bytes(skb, rroffset + sizeof(struct dnsrr),
-					   &event->anaddr[index][12], rdlength);
-			index++;
-		} else if (rrtype == DNS_TYPE_AAAA && rrclass == DNS_CLASS_IN &&
-			   rdlength == 16) {
-			// AAAA record contains an IPv6 address.
-			bpf_skb_load_bytes(skb, rroffset + sizeof(struct dnsrr),
-					   &event->anaddr[index][0], rdlength);
-			index++;
-		}
-		rroffset += sizeof(struct dnsrr) + rdlength;
-	}
-	return index;
-}
+	// Skip non-UDP packets
+	if (load_byte(skb, ETH_HLEN + offsetof(struct iphdr, protocol)) !=
+	    IPPROTO_UDP)
+		return 0;
 
-static __always_inline int output_dns_event(struct __sk_buff *skb,
-					    union dnsflags flags,
-					    __u32 name_len, __u16 ancount)
-{
-	__u32 zero = 0;
-	struct event_t *event = bpf_map_lookup_elem(&tmp_event, &zero);
-	if (!event)
+	sport = load_half(skb, ETH_HLEN + sizeof(struct iphdr) +
+				       offsetof(struct udphdr, source));
+	dport = load_half(skb, ETH_HLEN + sizeof(struct iphdr) +
+				       offsetof(struct udphdr, dest));
```

Comment:
I know this is the way we were doing it and you changed it in next commit but, I was wondering if it's safe to do this knowing that the IPv4 header doesn't have a fixed size. Shouldn't we always compute the layer 4 offset using the IHL field as you did in the next commit?

Reply:
As you mentioned, it's fixed by next commit. I wanted to keep the first commit as atomic as possible by only changing the things strictly needed. I'm still wondering if it's better to squash everything on a single commit and avoid the pain of having to divide it into two ones.

Reply:
I understand your reasoning behind dividing it into two commits, but it makes the review and (managing the comments for you) more difficult. For your mental sanity 🤣 , I think you should either create two PRs, one for parsing in user-space and another for supporting IPv6; or squash the commits.

Reply:
I'll drop the second commit from this PR. I can't handle it anymore.

Reply:
In any case, the original code wasn't handling it. I'll leave it as it was.